### DISCUSSION TITLE
Two Points to a unit vector

### DISCUSSION THREAD

**Discussion by Fengcheng-Pei** [e3nn/e3nn#513] [Q&A]
Hi everyone,

I'm new to e3nn and I'm trying to solve a problem where:
**Input**: two 3D points
**Output**: a unit vector
The mapping between the input and output is highly non-linear and complicated.

I first tried a standard MLP network, which worked very well on this task. However, when I switched to an e3nn-based network, the training loss remained very high. I experimented with both MSE loss and cosine similarity loss, and used a learning rate of 0.001 with a training set of 50,000 samples, but neither approach converged well.

I also tried **feature augmentation**, such as:

- mid-point of the two input points,

- direction vectors of each point,

- direction vector of the mid-point,

- distances of the two points and the mid-point.

Unfortunately, these additional features did not improve the results either.

Below is the architecture of my current e3nn model:

class E3NN_ra(nn.Module):
    def __init__(self, hidden_dim):
        super().__init__()

        self.irreps_in = o3.Irreps("1o+1o+1o+1o+1o+0e+0e+0e")
        self.irreps_hidden = o3.Irreps(f"{hidden_dim*2}x0e + {hidden_dim}x1e + {hidden_dim}x1o") #cross product!!!
        self.irreps_out = o3.Irreps("1x1o")

        # First layer: process the two input vectors
        self.tp1 = FullyConnectedTensorProduct(
            self.irreps_in,
            self.irreps_in,
            self.irreps_hidden
        )

        # Batch normalization
        self.bn1 = e3nn.nn.BatchNorm(self.irreps_hidden)

        # Non-linearity for scalars
        self.act1 = e3nn.nn.Activation(
            self.irreps_hidden,
            acts=[F.silu, None, None]  # silu for scalars, None for vectors
        )

        # Second layer
        self.tp2 = FullyConnectedTensorProduct(
            self.irreps_hidden,
            self.irreps_hidden,
            self.irreps_hidden
        )

        # Batch normalization
        self.bn2 = e3nn.nn.BatchNorm(self.irreps_hidden)

        # Non-linearity for scalars
        self.act2 = e3nn.nn.Activation(
            self.irreps_hidden,
            acts=[F.silu, None, None]  # silu for scalars, None for vectors and tensors
        )

        # Output layer
        self.tp_out = FullyConnectedTensorProduct(
            self.irreps_hidden,
            self.irreps_in,  # Use input irreps for tensor product
            self.irreps_out
        )

    def forward(self, point1, point2, D1, D2, M, L1, L2, LM):
        # Concatenate points for input
        x = torch.cat([point1, point2, D1, D2, M, L1, L2, LM], dim=1)

        # First layer
        h = self.tp1(x, x)
        h = self.act1(h)
        h = self.bn1(h)

        # Second layer
        h = self.tp2(h, h)
        h = self.act2(h)
        h = self.bn2(h)

        # Output layer
        output = self.tp_out(h, x)

        # Normalize to get unit vector
        output = F.normalize(output, p=2, dim=1)

        return output

Does anyone have suggestions on:
How to improve training for this kind of task using e3nn?

Any advice would be greatly appreciated. Thank you for your kind help!

**Comment by blondegeek** [comment]
Hi!

I suspect you are running into a problem of the problem not being
symmetrically well posed. Since we only operate on relative vectors between
points and you only have two points, by symmetry, the only non-zero vector
outputs that the model will be able to predict are along the axis formed by
the two points. I would need more information to debug further. Where are
you expecting these vectors to point relative to the two points? Is the
origin of your coordinate system meaningful? Note that if you are for
example trying to predict a point that lies perpendicular to this two point
axis, there are an infinite number of symmetrically equivalent predictions
that could be made. If you are trying to do something like this, you can
instead predict spherical harmonics that will give you the projection of
the distribution of possible vectors (even if you train on a single example
— symmetrically it is a distribution). We do this in this paper:
https://arxiv.org/abs/2311.16199

Best,
Tess

**Comment by blondegeek** [comment]
Ah, okay actually your network setup is actually a bit different from how
we normally act on points — typically we use a graph style neural network
to achieve translation equivariance in addition to rotation equivariance.
What I was saying holds in that case. In your case, you don’t have
translation equivariance because you are giving the raw coordinates as
input.

It is still possible you may be running into a symmetry issue, but I would
need more information to say. What are is the meaning of the three other 1o
features input to your network? Is there a pattern to the outputs the model
gives? There’s a chance you might also want to keep pseudoscalars 0o in
your hidden layers.

Best,
Tess

**Reply by Fengcheng-Pei** [reply]
Hi,

Thank you very much for your kind response. Actually, I am working on an optimization problem in communication engineering. In my setup:

- The **two points** represent the coordinates of two communication users.

- The **origin** corresponds to the position of the base station (an antenna array).

- The **output unit vector** represents the optimal steering direction of the linear antenna array.

The algorithm that computes the optimal antenna direction is highly complex, and I want to use deep learning to approximate this mapping. Importantly, when the user coordinates are rotated around the origin, the output antenna direction should rotate accordingly as well. 

In my input, I have five vectors: the coordinates of point1, the coordinates of point2, the unit direction of point1 as seen from the origin, the unit direction of point2 as seen from the origin, and the midpoint of the two points. I also have three scalar features: the distance of point1 from the origin, the distance of point2 from the origin, and the distance of the midpoint from the origin. Regarding the hidden layer irreps, I honestly have no clear idea, so I just chose some possible irreps arbitrarily.

Besides, there is some symmetry in my problem. For example, for my one-dimensional antenna array, if **r** is the optimal direction (the output unit vector), then -**r** is also optimal. However, I simply choose one of them as the supervised output. I used my e3nn structure to test a very simple problem: predicting the unit direction between two points, and it worked very well. However, I am not sure which structural aspects of my actual problem should be better exploited to adjust the e3nn architecture. Or should I use GNN with e3nn to do it?

Best regrads,
Fengcheng Pei

**Comment by blondegeek** [comment]
It’s possible you might need hidden pseudo scalars for cross product
information to propagate to an vector output (so you can make predictions
out of the plane) — something like 1o x 1o -> 1e, 1e x 0o -> 1o. You could
also try a graph conv net where the bulk of learnable parameters are in the
radial functions of a convolution so you could try a three point input
where you have things one hot encoded as origin (antenna) vs communication
users. Otherwise you might be facing more engineering type issues of
gradients, initializations, that sort of thing. But adding pseudoscalars to
the hidden would be the first thing I try since it’s easy.

**Comment by mariogeiger** [comment]
Hello,

This discussion is interesting:)

I would suggest trying with a GNN (as Tess was saying) and simply feed it 3 points in space. Decorate the points with a one-hot encoding (2x0e) to tell whether they are an antenna or a user. 

This would even add an extra symmetry of permutation of the users (if this is not a symmetry of your problem then you should distinguish between user 1 and user 2 by having an extra one hot channel).

**Comment by mitkotak** [comment]
Just sharing our group's current "engineered" example with all the normalization tricks incase you hit any roadblocks: https://github.com/prescient-design/e3tools/blob/main/examples/models/conv.py
